File: android/src/main/java/com/sendbird/calls/reactnative/CallsEvents.kt
package com.sendbird.calls.reactnative

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.sendbird.calls.reactnative.module.CallsModule
import com.sendbird.calls.reactnative.utils.CallsUtils

fun defaultType(type: String): String {
    return "${CallsEvents.EVENT_DEFAULT}.${type}"
}
fun directCallType(type: String): String {
    return "${CallsEvents.EVENT_DIRECT_CALL}.${type}"
}

fun groupCallType(type: String): String {
    return "${CallsEvents.EVENT_GROUP_CALL}.${type}"
}

class CallsEvents {
    companion object {
        /**
         * Event
         * EVENT_{EVENT}
         * */
        const val EVENT_DEFAULT = "sendbird.call.default"
        const val EVENT_DIRECT_CALL = "sendbird.call.direct"
        const val EVENT_GROUP_CALL = "sendbird.call.group"

        /**
         * Event Type
         * TYPE_{EVENT}_{TYPE}
         * */
        // Direct
        val TYPE_DEFAULT_ON_RINGING = defaultType("onRinging")
        val TYPE_DIRECT_CALL_ON_ESTABLISHED = directCallType("onEstablished")
        val TYPE_DIRECT_CALL_ON_CONNECTED = directCallType("onConnected")
        val TYPE_DIRECT_CALL_ON_RECONNECTING = directCallType("onReconnecting")
        val TYPE_DIRECT_CALL_ON_RECONNECTED = directCallType("onReconnected")
        val TYPE_DIRECT_CALL_ON_ENDED = directCallType("onEnded")
        val TYPE_DIRECT_CALL_ON_REMOTE_AUDIO_SETTINGS_CHANGED = directCallType("onRemoteAudioSettingsChanged")
        val TYPE_DIRECT_CALL_ON_REMOTE_VIDEO_SETTINGS_CHANGED = directCallType("onRemoteVideoSettingsChanged")
        val TYPE_DIRECT_CALL_ON_LOCAL_VIDEO_SETTINGS_CHANGED = directCallType("onLocalVideoSettingsChanged")
        val TYPE_DIRECT_CALL_ON_REMOTE_RECORDING_STATUS_CHANGED = directCallType("onRemoteRecordingStatusChanged")
        val TYPE_DIRECT_CALL_ON_AUDIO_DEVICE_CHANGED = directCallType("onAudioDeviceChanged")
        val TYPE_DIRECT_CALL_ON_CUSTOM_ITEMS_UPDATED = directCallType("onCustomItemsUpdated")
        val TYPE_DIRECT_CALL_ON_CUSTOM_ITEMS_DELETED = directCallType("onCustomItemsDeleted")
        val TYPE_DIRECT_CALL_ON_USER_HOLD_STATUS_CHANGED = directCallType("onUserHoldStatusChanged")
        // Group
        val TYPE_GROUP_CALL_ON_DELETED = groupCallType("onDeleted")
        val TYPE_GROUP_CALL_ON_ERROR = groupCallType("onError")
        val TYPE_GROUP_CALL_ON_REMOTE_PARTICIPANT_ENTERED = groupCallType("onRemoteParticipantEntered")
        val TYPE_GROUP_CALL_ON_REMOTE_PARTICIPANT_EXITED = groupCallType("onRemoteParticipantExited")
        val TYPE_GROUP_CALL_ON_REMOTE_PARTICIPANT_STREAM_STARTED = groupCallType("onRemoteParticipantStreamStarted")
        val TYPE_GROUP_CALL_ON_AUDIO_DEVICE_CHANGED = groupCallType("onAudioDeviceChanged")
        val TYPE_GROUP_CALL_ON_REMOTE_VIDEO_SETTINGS_CHANGED = groupCallType("onRemoteVideoSettingsChanged")
        val TYPE_GROUP_CALL_ON_REMOTE_AUDIO_SETTINGS_CHANGED = groupCallType("onRemoteAudioSettingsChanged")
        val TYPE_GROUP_CALL_ON_CUSTOM_ITEMS_UPDATED = groupCallType("onCustomItemsUpdated")
        val TYPE_GROUP_CALL_ON_CUSTOM_ITEMS_DELETED = groupCallType("onCustomItemsDeleted")

        fun sendEvent(reactContext: ReactContext, event: String, eventType: String, data: WritableMap) {
            Log.d(CallsModule.NAME, "[CallsEvents] sendEvent() $event++$eventType")

            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(event, Arguments.createMap().apply {
                    putString("eventType", eventType)
                    putMap("data", data)
                })
        }

        fun sendEvent(reactContext: ReactContext, event: String, eventType: String, data: List<*>) {
            Log.d(CallsModule.NAME, "[CallsEvents] sendEvent() $event++$eventType")

            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(event, Arguments.createMap().apply {
                    putString("eventType", eventType)
                    CallsUtils.insertMap(this, "data", data)
                })
        }

        fun sendEvent(reactContext: ReactContext, event: String, eventType: String, data: WritableMap, additionalData: Any) {
            Log.d(CallsModule.NAME, "[CallsEvents] sendEvent() $event++$eventType")

            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(event, Arguments.createMap().apply {
                    putString("eventType", eventType)
                    putMap("data", data)
                    CallsUtils.insertMap(this, "additionalData", additionalData)
                })
        }
    }
}

File: ios/Helpers/CallsEvents.swift
//
//  CallsEvents.swift
//  RNSendbirdCalls
//
//  Created by Airen Kang on 2022/05/07.
//  Copyright © 2022 Sendbird. All rights reserved.
//

import Foundation
import React

class CallsEvents {
    enum Event {
        case `default`(_ type: DefaultEventType)
        case directCall(_ type: DirectCallEventType)
        
        var name: String {
            switch self {
            case .`default`:
                return "sendbird.call.default"
            case .directCall:
                return "sendbird.call.direct"
            }
        }
        
        var type: String {
            switch self {
            case let .default(type: type):
                return "\(self.name).\(type.rawValue)"
            case let .directCall(type: type):
                return "\(self.name).\(type.rawValue)"
            }
        }
    }
    
    enum DefaultEventType: String, CaseIterable {
        case onRinging
    }
    
    enum DirectCallEventType: String, CaseIterable {
        case onEstablished
        case onConnected
        case onReconnecting
        case onReconnected
        case onEnded
        case onRemoteAudioSettingsChanged
        case onRemoteVideoSettingsChanged
        case onLocalVideoSettingsChanged
        case onRemoteRecordingStatusChanged
        case onAudioDeviceChanged
        case onCustomItemsUpdated
        case onCustomItemsDeleted
        case onUserHoldStatusChanged
    }
    
    static let shared = CallsEvents()
    
    var hasListeners = false
    var pendingEvents: [Dictionary<String, Any?>] = []
    var eventEmitter: RCTEventEmitter!
    var bridge: RCTBridge {
        get {
            self.eventEmitter.bridge
        }
    }
}

// MARK: RCTEventEmitter
extension CallsEvents {
    func startObserving() {
        hasListeners = true
        DispatchQueue.main.async {
            for event in self.pendingEvents {
                if let eventName = event["name"] as? CallsEvents.Event, let data = event["data"] as? Any {
                    self.sendEvent(eventName, data, event["additionalData"] ?? nil)
                }
            }
            self.pendingEvents.removeAll()
        }
    }
    
    func stopObserving() {
        hasListeners = false
    }
}

// MARK: Custom event sender
extension CallsEvents {
    func sendEvent(_ event: Event, _ data: Any) {
        if (hasListeners && eventEmitter != nil) {
            eventEmitter.sendEvent(withName: event.name, body: [
                "eventType": event.type,
                "data": data,
            ])
        } else {
            pendingEvents.append([
                "name": event,
                "data": data
            ])
        }
    }
    
    func sendEvent(_ event: Event, _ data: Any, _ additionalData: Any?) {
        if (hasListeners && eventEmitter != nil) {
            eventEmitter.sendEvent(withName: event.name, body: [
                "eventType": event.type,
                "data": data,
                "additionalData": additionalData ?? [:]
            ])
        } else {
            pendingEvents.append([
                "name": event,
                "data": data,
                "additionalData": additionalData
            ])
        }
    }
}
